### PR TITLE
sysupdate: Add support for optional features

### LINKED ISSUE #33343: Implement optional transfers in systemd-sysupdate
### Component

_No response_

### Is your feature request related to a problem? Please describe

Currently all transfers defined in `sysupdate.d` are taken into account and downloaded. This means we can't have optional transfers.

### Describe the solution you'd like

We want to have optional components of an OS that can be enabled/disabled by the user and acted upon by systemd-sysupdate.

The proposal is as follows:
* Have 2 directories considered by systemd-sysupdate: `sysupdate.d` and `sysregistry.d`
* Files in both directories are considered for ownership of currently installed stuff
    * that is `vacuum` will consider them when deleting old updates
* Files in `sysupdated.d` are downloaded when update is requested
* Files in `sysregistry.d` are not downloaded by default
    * These files can be symlinked into `sysupdate.d` to be considered for download
    * `vacuum` will delete things that are defined in `sysregistry.d` but not in `sysupdate.d` for the current version
    * 
* `systemd-sysupdate` and/or `updatectl` can get new verbs to enable/disable transfers by symlinking them from `sysregistry.d` to `sysupdate.d` / deleting the symlinks
    * Non-symlink files that are in `sysupdate.d` cannot be disabled (they aren't optional)

### Describe alternatives you've considered

_No response_

### The systemd version you checked that didn't have the feature you are asking for

256

### PR DESCRIPTION
**Pull request by AdrianVovk**
This adds a feature to systemd-sysupdate called "Optional Features", which makes it possible for distributions to group up certain transfers and let them be enabled/disabled by the system administrator.

TL;DR on how this works: The distro defines a `sysupdate.d/*.feature` file that defines a feature - the feature has an `Enabled=true/false` setting that lets the distro declare the default state of the feature. Then the system administrator can override the `Enabled` setting via a drop-in to toggle the feature on or off. Then transfers can associate themselves with a feature by setting `[Transfer] Features=<name of feature>`. If a transfer's associated features are all disabled, the transfer isn't downloaded. See the man pages introduced by this PR for more details.

Since the transfers associated with features are still in the same target as any other transfer, `@v` will resolve to the same version as the rest of the target, and updates to the target will only succeed if the optional transfer is successfully updated. In other words: enabled optional features become a part of their transfer.

Depends on:
- #33520
- #32363 
- #33570 
- #34079
- #33609 

Fixes #33343 and #33344

### CODE REVIEW

**Comment by calebccff** [comment]
For those of us interested in following along who haven't necessarily been keeping up with the rest of the systemd ecosystem, could you maybe write some PR description explaining what this does?

Would love to start following along with this.

**Comment by AdrianVovk** [comment]
@calebccff I've added a very high-level summary. There's also the updated docs in https://github.com/systemd/systemd/pull/33398/commits/d4ae918e64629f476e38ee389ce0d99dc3fa6f15 that shed light on the reasoning for this (specifically, the last paragraph I add to the docs in that commit).

File: src/sysupdate/sysupdate.c [hunk 1]
```diff
@@ -95,85 +115,152 @@ static Context *context_new(void) {
...
+                dirs = new0(char*, strv_length(l) + 1);
+                if (!dirs)
                         return log_oom();
 
-                STRV_FOREACH(i, l) {
+                STRV_FOREACH(dir, l) {
                         char *j;
 
-                        j = strjoin(*i, "sysupdate.", component, ".d");
+                        j = strjoin(*dir, "sysupdate", arg_component, ".d");
```
Comment (AdrianVovk) [inline src/sysupdate/sysupdate.c]: Whoops! lost the .

File: src/sysupdate/updatectl.c [hunk 2]
```diff
@@ -1151,6 +1437,7 @@ static int parse_argv(int argc, char *argv[]) {
                 { "host",      required_argument, NULL, 'H'             },
                 { "reboot",    no_argument,       NULL, ARG_REBOOT      },
                 { "offline",   no_argument,       NULL, ARG_OFFLINE     },
+                { "now",       no_argument,       NULL, ARG_OFFLINE     },
```
Comment (tomcoldrick-ct) [inline src/sysupdate/updatectl.c]: s/ARG_OFFLINE/ARG_NOW/

File: src/sysupdate/sysupdated.c [hunk 3]
```diff
@@ -75,6 +76,7 @@ typedef enum JobType {
         JOB_CHECK_NEW,
         JOB_UPDATE,
         JOB_VACUUM,
+        JOB_DESCRIBE_FEATURE,
```
Comment (AdrianVovk) [inline src/sysupdate/sysupdated.c:91]: Make sure I'm looking for all switch cases where these are used, and adding this case.

For example: job_method_cancel will panic the daemon if you try to cancel on of these jobs.

File: man/org.freedesktop.sysupdate1.xml [hunk 4]
```diff
@@ -274,6 +286,38 @@ node /org/freedesktop/sysupdate1/target/host {
       <varname>IMAGE_VERSION</varname> in <filename>/etc/os-release</filename>. If the target has no current
       version, the function will return an empty string.</para>
 
+      <para><function>ListFeatures()</function> returns a list of this target's optional features, by ID.
+      If the target has no optional features, the method returns an empty array.</para>
+
+      <para><function>DescribeFeature()</function> returns all known information about a given optional
```
Comment (AdrianVovk) [inline man/org.freedesktop.sysupdate1.xml]: Should use flags

File: man/org.freedesktop.sysupdate1.xml [hunk 5]
```diff
@@ -274,6 +286,38 @@ node /org/freedesktop/sysupdate1/target/host {
...
+      feature.
+      This method returns a JSON object with the following known keys:
+      <literal>name</literal> is a string that contains the feature's name,
```
Comment (AdrianVovk) [inline man/org.freedesktop.sysupdate1.xml]: Format as a nice table, like `Describe()` above

File: man/org.freedesktop.sysupdate1.xml (hunk 4 above)
Comment (AdrianVovk) [reply man/org.freedesktop.sysupdate1.xml]: Apparently in the code I read a variable that wasn't defined in the dbus. So no, shouldn't use flags. There are no flags to pass.

**Comment by yuwata** [comment]
Please rebase.

File: src/basic/alloc-util.h [hunk 6]
```diff
@@ -155,7 +155,10 @@ void* greedy_realloc_append(void **p, size_t *n_p, const void *from, size_t n_fr
         greedy_realloc0((void**) &(array), (need), sizeof((array)[0]))
 
 #define GREEDY_REALLOC_APPEND(array, n_array, from, n_from)             \
-        greedy_realloc_append((void**) &(array), (size_t*) &(n_array), (from), (n_from), sizeof((array)[0]))
+        ({                                                              \
+                const typeof(*(array)) *_from_ = (from);                \
+                greedy_realloc_append((void**) &(array), &(n_array), _from_, (n_from), sizeof((array)[0])); \
+        })
```
Comment (poettering) [inline src/basic/alloc-util.h]: might make sense to split out into separate mini PR we can merge quickly.

File: man/sysupdate.d.xml [hunk 7]
```diff
@@ -519,6 +529,39 @@
...
+        <citerefentry><refentrytitle>sysupdate.features</refentrytitle><manvolnum>5</manvolnum></citerefentry>
+        that this transfer belongs to, by name.
+        This option may be specified more than once, in which case the specified list of features is merged.
+        If the empty string is assigned to this option, the list is reset and all prior assignments will have
+        no effect.
+        For example: <literal>Features=foo bar</literal> specifies that the transfer belongs to
+        <literal>foo.feature</literal> and <literal>bar.feature</literal>.</para>
+
+        <para>If the list of features is empty, then this transfer is always used.
+        If this transfer belongs to more than one feature, then it will be used if any one of the listed
```
Comment (poettering) [inline man/sysupdate.d.xml]: I think `any` could use `<emphasize>` around it

File: man/sysupdate.d.xml [hunk 8]
```diff
@@ -869,7 +912,7 @@
       partition, and a unified kernel image to update as one. This example is an extension of the example
       discussed earlier in this man page.</para>
 
-      <para><programlisting># /usr/lib/sysupdate.d/50-verity.conf
+      <para><programlisting># /usr/lib/sysupdate.d/50-verity.transfer
 [Transfer]
```
Comment (poettering) [inline man/sysupdate.d.xml:916]: might have been nice to do the .conf → .transfer renaming in a separate commit. and adding .feature in a second commit.

but i understand this is quite involved splitting now, hence not asking for it.

File: man/sysupdate.features.xml [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+      <para>The developers of foobarOS can enable the <literal>devel</literal> feature on their workstations
+      by creating the following drop-in:</para>
+
+      <para><programlisting># /etc/sysupdate.d/devel.feature.d/enable.conf
+[Feature]
+Enabled=true
+</programlisting></para>
+    </example>
+
+
```
Comment (poettering) [inline man/sysupdate.features.xml]: double empty line

File: src/sysupdate/sysupdate-feature.c [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+#include "conf-parser.h"
+#include "hash-funcs.h"
+#include "path-util.h"
+#include "sysupdate-feature.h"
+#include "sysupdate.h"
+#include "web-util.h"
+
+static Feature *feature_free(Feature *f) {
+        free(f->id);
```
Comment (poettering) [inline src/sysupdate/sysupdate-feature.c:14]: please add `if (!f) return NULL;` at the top before dereffing `f`. We want that our destructors are generally happy with a NULL argument

File: src/sysupdate/sysupdate-feature.c [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        free(f->documentation);
+        free(f->appstream);
+
+        return mfree(f);
+}
+
+Feature *feature_new(void) {
+        Feature *f;
+
+        f = new0(Feature, 1);
```
Comment (poettering) [inline src/sysupdate/sysupdate-feature.c]: `new0` → `new`

(after all you use structured initialization right after, hence no need to zero-initialize things)

File: src/sysupdate/sysupdate-transfer.c [hunk 12]
```diff
@@ -466,18 +460,42 @@ static int config_parse_partition_flags(
...
-        assert(path);
 
         ConfigTableItem table[] = {
                 { "Transfer",    "MinVersion",              config_parse_min_version,          0, &t->min_version             },
                 { "Transfer",    "ProtectVersion",          config_parse_protect_version,      0, &t->protected_versions      },
                 { "Transfer",    "Verify",                  config_parse_bool,                 0, &t->verify                  },
                 { "Transfer",    "ChangeLog",               config_parse_url_specifiers,       0, &t->changelog               },
                 { "Transfer",    "AppStream",               config_parse_url_specifiers,       0, &t->appstream               },
+                { "Transfer",    "Features",                config_parse_strv,                 0, &features                   },
+                { "Transfer",    "RequisiteFeatures",       config_parse_strv,                 0, &req_features               },
```
Comment (poettering) [inline src/sysupdate/sysupdate-transfer.c]: hmm, i think we should store these two in Transfer as well. Yes, we only need them early on, but still, not a fan of such "optimizations". It# so little memory, and it makes the description more complete and debuggable. Hence, just add two fiels to Transfer for this.

File: src/sysupdate/sysupdate-transfer.c [hunk 13]
```diff
@@ -466,18 +460,42 @@ static int config_parse_partition_flags(
         return 0;
 }
 
-int transfer_read_definition(Transfer *t, const char *path) {
-        int r;
+static bool decide_if_enabled(Hashmap *known, char **regular, char **requisite) {
```
Comment (poettering) [inline src/sysupdate/sysupdate-transfer.c]: as suggested below, i think the two arrays should be part of the Transfer object, so that this function can then just take that as first argument. And maybe name the function transfer_decide_if_enabled() or so.

Comment (poettering) [inline src/sysupdate/sysupdate-transfer.c]: maybe rename `known` to `known_features`?

File: src/sysupdate/sysupdate-transfer.h [hunk 14]
```diff
@@ -15,12 +15,13 @@ typedef struct Transfer Transfer;
 #include "sysupdate.h"
 
 struct Transfer {
-        char *definition_path;
```
Comment (poettering) [inline src/sysupdate/sysupdate-transfer.h:18]: any reason you are dropping this? seems to be good to have for debugging, and doesn't cost much?

File: src/sysupdate/sysupdate.c [hunk 15]
```diff
@@ -773,15 +865,29 @@ static int context_vacuum(
...
                 else
-                        log_info("Removed no instances.");
+                        log_info("Found nothing to remove.");
         } else {
                 _cleanup_(sd_json_variant_unrefp) sd_json_variant *json = NULL;
 
-                r = sd_json_buildo(&json, SD_JSON_BUILD_PAIR_INTEGER("removed", count));
+                r = sd_json_buildo(&json,
+                                   SD_JSON_BUILD_PAIR_INTEGER("removed", count),
+                                   SD_JSON_BUILD_PAIR_UNSIGNED("disabled_transfers", disabled_count));
```
Comment (poettering) [inline src/sysupdate/sysupdate.c]: disabledTransfers

File: src/sysupdate/sysupdated.c [hunk 16]
```diff
@@ -1247,6 +1274,150 @@ static int target_method_get_appstream(sd_bus_message *msg, void *userdata, sd_b
         return sd_bus_send(NULL, reply, NULL);
 }
 
+
```
Comment (poettering) [inline src/sysupdate/sysupdated.c]: double empty line

File: src/sysupdate/sysupdated.c [hunk 17]
```diff
@@ -1247,6 +1274,150 @@ static int target_method_get_appstream(sd_bus_message *msg, void *userdata, sd_b
...
+        sd_json_variant *v;
+        uint64_t flags;
+        int r;
+
+        assert(msg);
+
+        r = sd_bus_message_read(msg, "t", &flags);
+        if (r < 0)
+                return r;
+
```
Comment (poettering) [inline src/sysupdate/sysupdated.c]: drpo this empty line, i.e. reading and validating should be kept close

File: src/sysupdate/sysupdated.c [hunk 18]
```diff
@@ -1247,6 +1274,150 @@ static int target_method_get_appstream(sd_bus_message *msg, void *userdata, sd_b
...
+        if (r == 0)
+                return 1; /* Will call us back */
+
+        /* We write to a file named 50-systemd-sysupdate-enabled.conf, because it's _very_ unlikely that
+         * a sysadmin would name their own config files that in this context. */
+        r = write_drop_in_format(SYSCONF_DIR "/sysupdate.d", feature, 50, "systemd-sysupdate-enabled",
+                                 "# Generated via org.freedesktop.sysupdate1 D-Bus interface\n\n"
+                                 "[Transfer]\n"
+                                 "Enabled=%s\n",
+                                 true_false(enabled));
```
Comment (poettering) [inline src/sysupdate/sysupdated.c]: i'd use yes_no()

File: src/sysupdate/sysupdated.c [hunk 19]
```diff
@@ -1247,6 +1274,150 @@ static int target_method_get_appstream(sd_bus_message *msg, void *userdata, sd_b
...
+        int enabled, r;
+
+        assert(msg);
+
+        if (t->class != TARGET_HOST)
+                return sd_bus_reply_method_errorf(msg,
+                                                  SD_BUS_ERROR_NOT_SUPPORTED,
+                                                  "For now, features can only be managed on the host system.");
+
+        r = sd_bus_message_read(msg, "sbt", &feature, &enabled, &flags);
```
Comment (poettering) [inline src/sysupdate/sysupdated.c]: hmm, i think it would make sense to have a mechanism to "reset" a feature to default state. i.e. maybe make the enabled thing a tristate where < 0 means "delete the dropin".

or add a separate method.

File: src/sysupdate/sysupdated.c [hunk 20]
```diff
@@ -1247,6 +1274,150 @@ static int target_method_get_appstream(sd_bus_message *msg, void *userdata, sd_b
...
+                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Flags must be 0");
+
+        if (!endswith(feature, ".feature")) {
+                feature_ext = strjoin(feature, ".feature");
+                if (!feature_ext)
+                        return -ENOMEM;
+                feature = feature_ext;
+        }
+
+        if (!filename_is_valid(feature))
```
Comment (poettering) [inline src/sysupdate/sysupdated.c]: i'd prefer if we'd have a proper feature_name_is_valid() helper somewhere, that ensures the feature name is not empty, and ascii only. and that when you add .feature or .feature.d you get a valid filename.

File: src/sysupdate/sysupdated.c [hunk 21]
```diff
@@ -1397,7 +1568,7 @@ static const sd_bus_vtable target_vtable[] = {
 
         SD_BUS_METHOD_WITH_ARGS("Vacuum",
                                 SD_BUS_NO_ARGS,
-                                SD_BUS_RESULT("u", count),
+                                SD_BUS_RESULT("u", instances, "u", disabled_transfers),
```
Comment (poettering) [inline src/sysupdate/sysupdated.c:1590]: i guess this is not really a compat break, given we haven#t released this yet hence ok

File: man/org.freedesktop.sysupdate1.xml [hunk 22]
```diff
@@ -489,9 +489,9 @@ node /org/freedesktop/sysupdate1/job/_1 {
 
       <para>The <varname>Id</varname> property exposes the numeric job ID of the job object.</para>
```
Comment (poettering) [inline man/org.freedesktop.sysupdate1.xml:492]: i'd probably still do this out of process. i think we might eventually want to hook more stuff into this, for example, automatically enable features based on the udev db, but if that's the case things that are "fast" right now, might eventually not be quite as fast anymore.

hence, doesn't matter, but given you already are doing this out-of-prcess, stick to it.

File: src/sysupdate/updatectl.c [hunk 23]
```diff
@@ -1165,27 +1217,259 @@ static int verb_vacuum(int argc, char **argv, void *userdata) {
...
+
+typedef struct Feature {
+        char *name;
+        char *description;
+        bool enabled;
+        char *documentation;
+        char **transfers;
+} Feature;
+
+static void feature_done(Feature *f) {
```
Comment (poettering) [inline src/sysupdate/updatectl.c:1235]: maybe do `assert(f);` as first thing for such a shallow destructor

**Comment by poettering** [comment]
looks pretty good to me, only minor stuff.

File: src/sysupdate/sysupdate-transfer.h (hunk 14 above)
Comment (AdrianVovk) [reply src/sysupdate/sysupdate-transfer.h:18]: It was unused, and I replaced it with an ID which is used. Seemed redundant

File: src/sysupdate/sysupdated.c [hunk 24]
```diff
@@ -1247,6 +1274,150 @@ static int target_method_get_appstream(sd_bus_message *msg, void *userdata, sd_b
...
+        if (r < 0)
+                return r;
+        if (r == 0)
+                return 1; /* Will call us back */
+
+        /* We write to a file named 50-systemd-sysupdate-enabled.conf, because it's _very_ unlikely that
+         * a sysadmin would name their own config files that in this context. */
+        r = write_drop_in_format(SYSCONF_DIR "/sysupdate.d", feature, 50, "systemd-sysupdate-enabled",
+                                 "# Generated via org.freedesktop.sysupdate1 D-Bus interface\n\n"
+                                 "[Transfer]\n"
```
Comment (AdrianVovk) [inline src/sysupdate/sysupdated.c]: Should be [Feature]?

File: src/sysupdate/sysupdated.c (hunk 21 above)
Comment (AdrianVovk) [reply src/sysupdate/sysupdated.c:1590]: yep. Part of the reason this is an RC blocker :)

File: src/sysupdate/sysupdated.c (hunk 20 above)
Comment (AdrianVovk) [reply src/sysupdate/sysupdated.c]: Presuably the "combined" filename is checked in the drop-in name mangling, right?

Not sure a helper is very useful since it's going to be used exactly once, here

File: src/basic/alloc-util.h (hunk 6 above)
Comment (AdrianVovk) [reply src/basic/alloc-util.h]: https://github.com/systemd/systemd/pull/34813

File: man/sysupdate.d.xml (hunk 8 above)
Comment (AdrianVovk) [reply man/sysupdate.d.xml:916]: yeah I agree it'd be better split. I tried to split it but it wasn't trivial (i.e. just `git add -p` during a rebase) so I didn't pursue it further

File: src/sysupdate/sysupdated.c (hunk 20 above)
Comment (poettering) [reply src/sysupdate/sysupdated.c]: still, we have that as something like a pattern: we have separate validator functions for validating such id strings.

**Comment by poettering** [comment]
CI is not happy:

```
1367s ../src/sysupdate/sysupdated.c: In function ‘target_method_set_feature_enabled’:
1367s ../src/sysupdate/sysupdated.c:1394:36: error: comparison is always true due to limited range of data type [-Werror=type-limits]
1367s  1394 |                 "enabled", enabled >= 0 ? true_false(enabled) : "unset",
1367s       |                                    ^~
1367s ../src/sysupdate/sysupdated.c:1410:21: error: comparison is always false due to limited range of data type [-Werror=type-limits]
1367s  1410 |         if (enabled < 0) { /* Reset -> delete the drop-in file */
1367s       |                     ^
1367s cc1: all warnings being treated as errors
```

File: src/sysupdate/sysupdated.c [hunk 25]
```diff
@@ -1247,6 +1276,159 @@ static int target_method_get_appstream(sd_bus_message *msg, void *userdata, sd_b
...
+
+        return 1;
+}
+
+static int target_method_set_feature_enabled(sd_bus_message *msg, void *userdata, sd_bus_error *error) {
+        _cleanup_free_ char *feature_ext = NULL;
+        Target *t = ASSERT_PTR(userdata);
+        const char *feature;
+        uint64_t flags;
+        char enabled;
```
Comment (poettering) [inline src/sysupdate/sysupdated.c]: char?

Comment (poettering) [reply src/sysupdate/sysupdated.c]: "char" can be signed or unsigned (which is where the CI issue comes from).

D-Bus doesn't know signed 8bit integers, it only knows "bytes" which are unsigned.

i'd just us int32_t for this. yeah, a lot of unused bits, but who cares, and internally we use `int` for tristate values anyway, and C99 `bool` typically ends up being `int` as well.

File: src/sysupdate/sysupdated.c [hunk 26]
```diff
@@ -1413,6 +1595,24 @@ static const sd_bus_vtable target_vtable[] = {
...
+                                SD_BUS_VTABLE_UNPRIVILEGED),
+
+        SD_BUS_METHOD_WITH_ARGS("DescribeFeature",
+                                SD_BUS_ARGS("s", feature, "t", flags),
+                                SD_BUS_RESULT("s", json),
+                                target_method_describe_feature,
+                                SD_BUS_VTABLE_UNPRIVILEGED),
+
+        SD_BUS_METHOD_WITH_ARGS("SetFeatureEnabled",
+                                SD_BUS_ARGS("s", feature, "y", enabled, "t", flags),
```
Comment (poettering) [inline src/sysupdate/sysupdated.c]: "y" → "i", in context of the above

File: src/sysupdate/sysupdate-transfer.c [hunk 27]
```diff
@@ -466,18 +460,40 @@ static int config_parse_partition_flags(
         return 0;
 }
 
-int transfer_read_definition(Transfer *t, const char *path) {
-        int r;
+static bool transfer_decide_if_enabled(Transfer *t, Hashmap *known_features) {
```
Comment (poettering) [inline src/sysupdate/sysupdate-transfer.c:468]: maybe `assert(t)` first

File: src/sysupdate/sysupdate.c [hunk 28]
```diff
@@ -1201,6 +1201,142 @@ static int verb_list(int argc, char **argv, void *userdata) {
...
+                        if (r < 0)
+                                return table_log_add_error(r);
+                }
+
+                if (f->documentation) {
+                        r = table_add_many(table,
+                                           TABLE_FIELD, "Documentation",
+                                           TABLE_STRING, f->documentation,
+                                           TABLE_SET_URL, f->documentation,
+                                           TABLE_SET_JSON_FIELD_NAME, "documentation_url");
```
Comment (poettering) [inline src/sysupdate/sysupdate.c]: "documentationUrl"

File: src/sysupdate/sysupdate.c [hunk 29]
```diff
@@ -1201,6 +1201,142 @@ static int verb_list(int argc, char **argv, void *userdata) {
...
+                        if (r < 0)
+                                return table_log_add_error(r);
+                }
+
+                if (f->appstream) {
+                        r = table_add_many(table,
+                                           TABLE_FIELD, "AppStream",
+                                           TABLE_STRING, f->appstream,
+                                           TABLE_SET_URL, f->appstream,
+                                           TABLE_SET_JSON_FIELD_NAME, "appstream_url");
```
Comment (poettering) [inline src/sysupdate/sysupdate.c]: appstreamUrl?

File: src/sysupdate/sysupdated.c [hunk 30]
```diff
@@ -1247,6 +1276,159 @@ static int target_method_get_appstream(sd_bus_message *msg, void *userdata, sd_b
...
+        int r;
+
+        assert(msg);
+
+        if (t->class != TARGET_HOST)
+                return sd_bus_reply_method_errorf(msg,
+                                                  SD_BUS_ERROR_NOT_SUPPORTED,
+                                                  "For now, features can only be managed on the host system.");
+
+        r = sd_bus_message_read(msg, "syt", &feature, &enabled, &flags);
```
Comment (poettering) [inline src/sysupdate/sysupdated.c]: syt → sit

File: src/basic/fs-util.h [hunk 31]
```diff
@@ -60,7 +60,10 @@ static inline int touch(const char *path) {
         return touch_file(path, false, USEC_INFINITY, UID_INVALID, GID_INVALID, MODE_INVALID);
 }
 
-int symlink_idempotent(const char *from, const char *to, bool make_relative);
+int symlinkat_idempotent(const char *from, int atfd, const char *to, bool make_relative);
+static inline int symlink_idempotent(const char *from, const char *to, bool make_relative) {
+        return symlinkat_idempotent(from, AT_FDCWD, to, make_relative);
+}
```
Comment (poettering) [inline src/basic/fs-util.h:66]: ideally split out into separate commit

File: src/sysupdate/updatectl.c [hunk 32]
```diff
@@ -1165,27 +1217,260 @@ static int verb_vacuum(int argc, char **argv, void *userdata) {
...
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        _cleanup_(feature_done) Feature f = {};
+        char *json;
+        int r;
+
+        static const sd_json_dispatch_field dispatch_table[] = {
+                { "name",              SD_JSON_VARIANT_STRING,  sd_json_dispatch_string,  offsetof(Feature, name),          SD_JSON_MANDATORY },
+                { "description",       SD_JSON_VARIANT_STRING,  sd_json_dispatch_string,  offsetof(Feature, description),   0                 },
+                { "enabled",           SD_JSON_VARIANT_BOOLEAN, sd_json_dispatch_stdbool, offsetof(Feature, enabled),       SD_JSON_MANDATORY },
+                { "documentation_url", SD_JSON_VARIANT_STRING,  sd_json_dispatch_string,  offsetof(Feature, documentation), 0                 },
```
Comment (poettering) [inline src/sysupdate/updatectl.c]: documentationUrl

File: man/org.freedesktop.sysupdate1.xml [hunk 33]
```diff
@@ -273,6 +288,68 @@ node /org/freedesktop/sysupdate1/target/host {
...
+          this feature</para></listitem>
+        </varlistentry>
+
+        <varlistentry>
+          <term><literal>enabled</literal></term>
+          <listitem><para>A boolean indicating whether this feature is enabled.</para></listitem>
+        </varlistentry>
+
+        <varlistentry>
+          <term><literal>documentation_url</literal></term>
```
Comment (poettering) [inline man/org.freedesktop.sysupdate1.xml]: I think we should prefer camelCase over snake_case for new JSON objects we introduce

**Comment by poettering** [comment]
looks all good, except for the "char" thing, please fix to "int32_t".

Otherwise just nitpicks about naming.

**Comment by AdrianVovk** [comment]
:tada: :tada: :tada: Thanks!

**Comment by pothos** [comment]
@AdrianVovk thanks for doing this work :) I'm happy that it also cleans any leftovers up :+1:
